src/index/IndexImpl.h

```diff
@@ -0,0 +1,731 @@
+
+  const string& getTextName() const { return _textMeta.getName(); }
+
+  const string& getKbName() const { return _PSO.metaData().getName(); }
+
+  size_t getNofTriples() const { return _PSO.metaData().getNofTriples(); }
+
+  size_t getNofTextRecords() const { return _textMeta.getNofTextRecords(); }
+  size_t getNofWordPostings() const { return _textMeta.getNofWordPostings(); }
+  size_t getNofEntityPostings() const {
+    return _textMeta.getNofEntityPostings();
+  }
+
+  size_t getNofSubjects() const {
+    if (hasAllPermutations()) {
+      return _SPO.metaData().getNofDistinctC1();
+    } else {
+      AD_THROW(ad_semsearch::Exception::CHECK_FAILED,
+               "Can only get # distinct subjects if all 6 permutations "
+               "have been registered on sever start (and index build time) "
+               "with the -a option.")
+    }
+  }
+
+  size_t getNofObjects() const {
+    if (hasAllPermutations()) {
+      return _OSP.metaData().getNofDistinctC1();
+    } else {
+      AD_THROW(ad_semsearch::Exception::CHECK_FAILED,
+               "Can only get # distinct subjects if all 6 permutations "
+               "have been registered on sever start (and index build time) "
+               "with the -a option.")
+    }
+  }
+
+  size_t getNofPredicates() const { return _PSO.metaData().getNofDistinctC1(); }
+
+  bool hasAllPermutations() const { return SPO()._file.isOpen(); }
+
+ private:
+  // _____________________________________________________________________________
+  template <class PermutationImplementation>
+  vector<float> getMultiplicitiesImpl(const PermutationImplementation& p,
+                                      const string& key) const;
+
+  // ___________________________________________________________________
+  template <class PermutationImplementation>
+  vector<float> getMultiplicitiesImpl(const PermutationImplementation& p) const;
+
+ public:
+  vector<float> getMultiplicities(const string& key, const Permutation p) const;
+  vector<float> getMultiplicities(const Permutation p) const;
+
+  /**
+   * @brief Perform a scan for one key i.e. retrieve all YZ from the XYZ
+   * permutation for a specific key value of X
+   * @tparam PermutationImplementation The permutations IndexImpl::POS()... have
+   * different types
+   * @param key The key (in Id space) for which to search, e.g. fixed value for
+   * O in OSP permutation.
+   * @param result The Id table to which we will write. Must have 2 columns.
+   * @param p The Permutation to use (in particularly POS(), SOP,... members of
+   * IndexImpl class).
+   */
+  void scan(Id key, IdTable* result, const Permutation& p,
+            ad_utility::SharedConcurrentTimeoutTimer timer = nullptr) const;
+
+ private:
+  // _____________________________________________________________________
+  template <PermutationConcept P>
+  void scanImpl(Id key, IdTable* result, const P& p,
+                ad_utility::SharedConcurrentTimeoutTimer timer = nullptr) const;
+
```

Comment:
This pattern is new.
The `scan` is the external interface (using a simple enum for the `Permutation`,
and the `scanImpl` does the actual work and gets a `PermutationConcept` aka `PermutationImpl::PermutationImpl<SomeTypes>`.
See the .cpp file for the details.

( there are two more `scan` overloads below with the same pattern.